File: front/lib/front/models/billing/invoice.ex
defmodule Front.Models.Billing.Invoice do
  alias InternalApi.Billing.Invoice, as: GrpcInvoice
  alias __MODULE__

  defstruct [:name, :total, :url]

  @type t :: %Invoice{
          name: String.t(),
          total: String.t(),
          url: String.t()
        }

  @spec new(Enum.t()) :: t()
  def new(params \\ %{}), do: struct(Invoice, params)

  @spec from_grpc(GrpcInvoice.t()) :: t()
  def from_grpc(invoice = %GrpcInvoice{}) do
    new(
      name: invoice.display_name,
      total: invoice.total_bill,
      url: invoice.pdf_download_url
    )
  end
end

File: front/assets/js/billing/types/spendings/invoice.ts
export class Invoice {
  name: string;
  url: string;
  total: string;

  static fromJSON(json: any): Invoice {
    const i = new Invoice();

    i.name = json.name as string;
    i.url = json.url as string;
    i.total = json.total as string;

    return i;
  }
}
